test/test_foreach.py

```diff
@@ -981,6 +983,34 @@ def inplace_func(*tensorlist):
                     check_batched_grad=False,
                 )
 
+    @unittest.skipIf(not (torch.cuda.is_available() and torch.cuda.device_count() > 1), "requires multiple GPUs")
+    def test_tensors_grouping(self):
+        num_tensors_per_list = 10
+        num_devices = torch.cuda.device_count()
+        dtypes = (torch.float16, torch.float32, torch.float64)
+        list1 = [
+            torch.tensor(
+                i,
+                device=torch.device("cuda", random.randint(0, num_devices - 1)),
+                dtype=dtypes[random.randint(0, 2)],
+            ) for i in range(num_tensors_per_list)
+        ]
+        list2 = [None for _ in list1]
+        list3 = [torch.rand_like(t) for t in list1]
+        nested_tensorlists = [list1, list2, list3]
+        grouped_tensors = torch.utils._foreach_utils._group_tensors_by_device_and_dtype(nested_tensorlists, with_indices=True)
+        num_tensors_seen = 0
+        for (device, dtype), ([l1, l2, l3], indices) in grouped_tensors.items():
+            self.assertEquals(l2, [None])
```

Comment:
Is it true that l2 will always only have one None?

Reply:
ah, it's up to the number of available devices. The check of your suggestion below would be simpler and correct